Add support for Atmel SAML21, SAMR34, SAMR35 parts

attie-argentum opened the pull request:
This PR implements support for SAML21, SAMR34 and SAMR35 parts.

I've named the SAMR34 / SAMR35 parts `samr34` in the sources, because `samr3x` would not be correct here (see [SAMR30](https://www.microchip.com/en-us/products/wireless-connectivity/sub-ghz/sam-r30) which is a SAML21 with an AT86RF212B / 802.15.4 radio), and SAMR35 is just "_SAMR34 without USB_".

The SAMR34 and SAMR35 parts are fundamentally a SAML21 with a LoRa radio built in.
I will not be using the LoRa radio for LoRa in my application (I'll be using FSK), but I have run the LoRa send/receive samples between two boards to verify operation.

## Dependencies

This PR depends on:
- zephyrproject-rtos/hal_atmel#17
- #41319
- #43731

attie-argentum committed:
soc: atmel_sam0: Add support for SAML21 parts

Implement support for SAML21 parts, tested on a SAML21 Xplained Pro dev
board. Confirmed operational peripherals: UART, GPIO, PWM, ADC.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
boards: arm: atsaml21: Add SAML21 Xplained board

Implement support for SAML21 Xplained Pro dev board.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

carlescufi commented:
> ## Dependencies
> This PR depends on [zephyrproject-rtos/hal_atmel#17](https://github.com/zephyrproject-rtos/hal_atmel/pull/17).

Can you please update west.yml in this PR to point to the PR above?

attie-argentum commented:
> Can you please update west.yml in this PR to point to the PR above?
What a good idea...

stephanosio commented on west.yml:
```diff
@@ -57,7 +57,7 @@ manifest:
         - hal
     - name: hal_atmel
-      revision: 9f78f520f6cbb997e5b44fe8ab17dd5bf2448095
+      revision: e5d2a27c9c1146916e6853c74645663c23be17bc
```
Please use `pull/17/head`.

attie-argentum replied:
... will that magically resolve to a commit ID once the PR is merged? (or will we require a second PR here once that one is in place?)

nandojve commented:
Hi @attie-argentum ,

Thank you for your contribution!

I think there are three things in this PR:

 - introduce a new SoC
 - improve ADC
 - fix pinctrl

My preference is split PR in two for organization purposes.

In order to add a newer SoC the preference is have a single commit that update `hal_atmel` using `pull/17/head`. It should be first commit in the series. The `pull/XX/head` signal PR must not be merged because it handles `DNM` (Do Not Merge) automatically. You can keep the pinctrl fix here or create another PR (trivial) that could be merged faster.

ADC is a general improvement and don't need to be part of a newer SoC series as you, in fact, fixed ADC before add the SoC.

The split can help to speed up acceptance.

attie-argentum commented:
@stephanosio, @nandojve - thanks for your comments.

As requested, I have split this into three PRs:

- #41308 (this PR)
- #41319 (this depends on it)
- #41320 (standalone)

attie-argentum commented:
Apologies for the multiple force-pushes... I believe the PR is now in good shape.

- The ADC unit test should be resolved once #41319 is merged
- The async UART test fails on SAMD21 boards too - it seems that DMA support for the SERCOM peripherals is missing, and I don't intend to look into this further.

attie-argentum committed:
soc: atmel_sam0: Add support for SAMR34 and SAMR35 parts

Implement support for SAMR34 and SAMR35 parts, tested on a SAMR34
Xplained Pro dev board. Confirmed operational peripherals: UART, GPIO,
PWM, ADC.

NOTE: The SAMR34 datasheet lists NVIC line 20 for the ADC's interrupt,
but the SAML21 datasheet gives line 22. It appears that the SAMR34 sheet
is incorrect (the ADC doesn't function with line 20), and as such, I have
also omitted the adjustments for TC4 and AC...

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
boards: arm: atsamr34: Add SAMR34 Xplained board

Implement support for SAMR34 Xplained Pro dev board.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented:
I've just got into a good position with the "_real_" project, and found that for the ASF integration, the SAMR34 and SAMR35 series need to be split.

nandojve commented on soc/arm/atmel_sam0/saml21/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,31 @@
+config NUM_IRQS
+	default 29
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 48000000
```
https://github.com/zephyrproject-rtos/zephyr/blob/44462723f59ddcc0786af0b5406122476c2addf9/soc/riscv/riscv-privilege/gd32vf103/Kconfig.defconfig.gd32vf103#L12-L13

nandojve commented on soc/arm/atmel_sam0/samr34/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,22 @@
+config NUM_IRQS
+	default 23
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 48000000
```
Try pick everything you can from devicetree as this https://github.com/zephyrproject-rtos/zephyr/blob/44462723f59ddcc0786af0b5406122476c2addf9/soc/riscv/riscv-privilege/gd32vf103/Kconfig.defconfig.gd32vf103#L12-L13

nandojve commented on soc/arm/atmel_sam0/samr34/soc.h:
```diff
@@ -0,0 +1,62 @@
+
+#include <zephyr/types.h>
+
+/* Add include for DTS generated information */
+#include <devicetree.h>
```
Please check #41544

nandojve commented on soc/arm/atmel_sam0/common/soc_samr34.c:
```diff
@@ -0,0 +1,43 @@
+#include <drivers/pinmux.h>
+#include <soc.h>
+
+static int soc_pinmux_init(const struct device *dev)
+{
+	const struct device *muxb = DEVICE_DT_GET(DT_NODELABEL(pinmux_b));
+	const struct device *muxc = DEVICE_DT_GET(DT_NODELABEL(pinmux_c));
```
Pinmux driver is in EOL. This means that it should not be encouraged to be at SoC directory in my view.
Please, keep at board for now. As soon SPI driver get pinctrl capabilities all spi pinmux will be dropped, see #39740

attie-argentum replied:
> Please, keep at board for now.

Even though this is SoC specific, not board specific?

> As soon SPI driver get pinctrl capabilities all spi pinmux will be dropped

SPI supporting pinctrl doesn't fully resolve the issue here... when the internal radio is not used, it must be held in reset with `#CS` high to avoid unpredictable radio behaviour and excessive power usage - this is called out explicitly in the datasheet.

I'd be quite uncomfortable putting this code on the board layer due to this...

Are we able to get similar (raw GPIO) functionality from `pinctrl.h`? A quick look suggests no...

nandojve reviewed (changes requested):
Hello @attie-argentum ,

I came back from a period of vacation and now I can continue to review your PR.
I pointed some parts that require few attention. Try replicate for all SoCs, as much you can, when applicable.

attie-argentum commented on boards/arm/atsamr34_xpro/atsamr34_xpro.dts:
```diff
@@ -0,0 +1,109 @@
+&sercom4 {
+	status = "okay";
+};
+&lora {
+	tcxo-power-gpios = <&porta 9 GPIO_ACTIVE_HIGH>;  /* TCXO_PWR */
+	tcxo-power-startup-delay-ms = <5>;
+	rfi-enable-gpios = <&porta 13 GPIO_ACTIVE_HIGH>; /* BAND_SEL */
+	rfo-enable-gpios = <&porta 13 GPIO_ACTIVE_HIGH>; /* BAND_SEL */
+};
```
This pin mapping / configuration is board-specific.

attie-argentum commented:
Thanks @nandojve - I'll try to get to this in the next few days... it's currently blocked by zephyrproject-rtos/hal_atmel#17 and #41319 anyway, so no rush.

attie-argentum committed:
soc: atmel_sam0: The ADC reference is locked while enabled for SAML21

For SAML21-based parts, the REFCTRL register is locked while the ADC is
enabled. Permit some parts to declare that they need the ADC to be
disabled before modifying REFCTRL.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented on samples/drivers/adc/boards/atsaml21_xpro.overlay:
```diff
@@ -0,0 +1,12 @@
+ * Copyright (c) 2021 Argentum Systems Ltd.
+ */
+
+/ {
+	zephyr,user {
+		/* EXT-1, pin 3 ADC(+) */
+		io-channels = <&adc 13>;
+	};
+};
```
The SAM0 ADC driver currently has no support for DMA, but I have added the node to the SAML21 device tree.

attie-argentum commented on dts/arm/atmel/samr34.dtsi:
```diff
@@ -0,0 +1,67 @@
+				#atmel,pin-cells = <2>;
+			};
+
+			/* Internal SPI link to SX127x radio */
+			DT_ATMEL_PORT(sercom4, pad2, b, 30, f, pinmux-enable); /* --> COPI */
+			DT_ATMEL_PORT(sercom4, pad1, b, 31, f, pinmux-enable); /* --> #CS */
+			DT_ATMEL_PORT(sercom4, pad3, c, 18, f, pinmux-enable); /* --> SCK */
+			DT_ATMEL_PORT(sercom4, pad0, c, 19, f, pinmux-enable); /* <-- CIPO */
```
> No neeed to describe here. Please, add as normal pinctrl pins
Done

> You can add a newer file like samr2x-pinctrl.dtsi
I don't understand, can you expand please?

attie-argentum replied:
I've added a `samr3x-pinctrl.dtsi`, is that what you were asking for?

attie-argentum commented:
@nandojve - I've made some of the requested changes, and have some queries... please review at let me know your thoughts.

gmarull commented on dts/arm/atmel/saml2x-pinctrl.dtsi:
```diff
@@ -0,0 +1,82 @@
+/ {
+	soc {
+		pinctrl@40002800 {
+			/* instance, signal, port, pin, peripheral [, flags] */
+			DT_ATMEL_PORT(sercom0, pad0, a,  4, d, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad1, a,  5, d, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad2, a,  6, d, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad3, a,  7, d, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad0, a,  8, c, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad1, a,  9, c, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad2, a, 10, c, pinmux-enable);
+			DT_ATMEL_PORT(sercom0, pad3, a, 11, c, pinmux-enable);
```
gentle reminder: all platforms should start migrating to pinctrl, ref https://github.com/zephyrproject-rtos/zephyr/issues/39740. The format used here (macros) should not be used.

nandojve replied:
@gmarull , this platform was there before this rule be created and, as far I remember, this could stay as is.

gmarull replied:
The Atmel pinmux driver came before pinctrl, so this topic was not discussed at that time. During pinctrl implementation it was concluded that such "helpers" were not a good idea because:

1. They hide Devicetree language
2. Could make external tooling (e.g. editor extensions) more difficult to implement

So, I'd prefer not to have yet another flavor of Devicetree representation unless it is strictly needed.

nandojve replied:
@gmarull , As far I remember, it was decided that vendor is free to choose about devicetree representation related to pinctrl. This is not a new flavor, it was the first pinctrl version to be in the tree.

Zephyr still on migrate phase and I think it is not appropriated to block something that is already in the tree.

On my view, I think most important is focus to drop pinmux drivers ASAP and after that work on a migration path.

gmarull replied:
So this case is not about changing DT operation mode really (boards will not require any update). It's about not using macros that expand to DT nodes.

Btw, the current format can be easily converted using a Python script like that:

```python
import re
import sys

output = ""
with open(sys.argv[1], "r") as f:
    for line in f:
        m = re.match(
            r"(\s+)DT_ATMEL_PORT\((.*),\s+(.*),\s+(.*),\s+(.*),\s+(.*),\s+(.*)\);", line
        )
        if not m:
            output += line
            continue

        args = "\n".join(
            (f"{m.group(1)}\t{arg.strip()};" for arg in m.group(7).split(","))
        )

        output += f"{m.group(1)}p{m.group(4)}{m.group(5)}{m.group(6)}_{m.group(2)}_{m.group(3)} {{\n"
        output += f"{m.group(1)}\tatmel,pins = <&port{m.group(4)} {m.group(5)} PERIPH_{m.group(6)}>;\n"
        output += args
        output += f"\n{m.group(1)}}};\n\n"

with open(sys.argv[1], "w") as f:
    f.write(output)
```

which converts content like:
```
			DT_ATMEL_PORT(sercom0, pad0, a,  4, d, pinmux-enable);
			DT_ATMEL_PORT(sercom0, pad1, a,  5, d, pinmux-enable);
```

to:

```
			pa4d_sercom0_pad0 {
				atmel,pins = <&porta 4 PERIPH_d>;
				pinmux-enable;
			};

			pa5d_sercom0_pad1 {
				atmel,pins = <&porta 5 PERIPH_d>;
				pinmux-enable;
			};
```

so the effort to align with pinctrl is really low.

nandojve replied:
No, what you suggests is not a final solution and any new pin to be added will imply in a lot of more effort. Zephyr and Devicetree itself are constructed above macro expansion. Here uses same principle, so I don't see any motive to not use it. There are pros/cons on each approach and the macro still better for Atmel use cases.

I prefer keep macro to define a pin until have a final solution, which will be similar to how STM/GD have been defining pins.

attie-argentum replied:
Are we agreeing to continue using `DT_ATMEL_PORT()`? If no, then #42512 (and a lot of other SAM0 sources) will require some rework too...

gmarull commented on boards/arm/atsaml21_xpro/pinmux.c:
```diff
@@ -0,0 +1,79 @@
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(3, atmel_sam0_spi) && CONFIG_SPI_SAM0)
+#warning Pin mapping may not be configured
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(4, atmel_sam0_spi) && CONFIG_SPI_SAM0)
+#warning Pin mapping may not be configured
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(5, atmel_sam0_spi) && CONFIG_SPI_SAM0)
+	/* SPI SERCOM5 on MISO=PB16/pad 0, MOSI=PB22/pad 2, SCK=PB23/pad 3 */
+	pinmux_pin_set(muxb, 16, PINMUX_FUNC_C);
+	pinmux_pin_set(muxb, 22, PINMUX_FUNC_D);
+	pinmux_pin_set(muxb, 23, PINMUX_FUNC_D);
+#endif
+
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(0, atmel_sam0_i2c) && CONFIG_I2C_SAM0)
+#warning Pin mapping may not be configured
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(1, atmel_sam0_i2c) && CONFIG_I2C_SAM0)
+#warning Pin mapping may not be configured
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(2, atmel_sam0_i2c) && CONFIG_I2C_SAM0)
+	/* SERCOM2 on SDA=PA08, SCL=PA09 */
+	pinmux_pin_set(muxa, 8, PINMUX_FUNC_D);
+	pinmux_pin_set(muxa, 9, PINMUX_FUNC_D);
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(3, atmel_sam0_i2c) && CONFIG_I2C_SAM0)
+#warning Pin mapping may not be configured
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(4, atmel_sam0_i2c) && CONFIG_I2C_SAM0)
+#warning Pin mapping may not be configured
+#endif
+#if (ATMEL_SAM0_DT_SERCOM_CHECK(5, atmel_sam0_i2c) && CONFIG_I2C_SAM0)
+#warning Pin mapping may not be configured
+#endif
+
+#if (ATMEL_SAM0_DT_TCC_CHECK(0, atmel_sam0_tcc_pwm) && CONFIG_PWM_SAM0_TCC)
+	/* TCC0 on WO4=PB10 - note WO4 is the inverse of WO0 */
+	pinmux_pin_set(muxb, 10, PINMUX_FUNC_F);
+#endif
+
+	return 0;
```
Please, implement a pinctrl driver. We should not be adding new code like this.

carlescufi replied:
Take a look at https://docs.zephyrproject.org/latest/guides/pinctrl/index.html

nandojve replied:
Hi @attie-argentum ,

I agreed with @gmarull in the pinmux topic and this file should be excluded.

The sam0 already have pinctrl driver implemented and serial driver already use it. The path is implement same approach for SPI, I2C and TCC.

attie-argentum replied:
I've implemented pinctrl in #42512, and revised this PR to make use of it... TCC pin muxing is still TODO...

attie-argentum commented:
I've now reworked this to use the pinctrl support implemented in #42512, and rebased on main.

attie-argentum commented:
Can anyone identify what's going wrong with `tests/drivers/uart/uart_async_api/drivers.uart.uart_async_api.rtt`? (for both SAML21 XPro and SAMR34 XPro)

I suspect the littlefs failures are not related to this PR, as they're seen in #42512 too.

gmarull commented:
> Please could I get some clarity on what is being requested on this PR? (primarily from @gmarull and @nandojve)
> 
> * Re changes to `dts/arm/atmel/saml2x-pinctrl.dtsi`: are we sticking with the `DT_ATMEL_PORT()` macro? If no, do all other SoCs that use the macro need to be revised too? (will need to be a separate PR, as/when time permits)
> * Re pinmux vs pinctrl, I proposed [Add Atmel SAM0 pinctrl for SPI, I2C & TCC #42512](https://github.com/zephyrproject-rtos/zephyr/pull/42512) that would get us moving in the correct direction, but this appears to be rejected in its current state (needs to be a "proper" pinctrl driver)... For the moment, I can unwind the changes to this PR to [`saml21-no-pinctrl`](https://github.com/argentum-systems/zephyr/tree/saml21-no-pinctrl?rgh-link-date=2022-02-11T12%3A40%3A39Z), but that would require the `pinmux.c` files that we are trying to avoid adding...
> * I don't have time right now to implement the "proper" pinctrl driver (maybe in a couple of weeks)... given the restrictions, my view is that falling inline with existing SoCs in this family is the most sensible approach.
> 
> Thoughts? What's the path to getting this merged?

So what is being requested is basically a new pinctrl driver, instead of continuing extending custom pinmux or board-level pin control, now that there's an API to solve this. You can find useful material here:

- https://docs.zephyrproject.org/latest/guides/pinctrl/index.html
- A few existing implementations have already been merged, you can check https://github.com/zephyrproject-rtos/zephyr/issues/39740 (also contains roadmap)

There's a `#pinctrl` channel in case you have questions, too.

attie-argentum commented:
> So what is being requested is basically a new pinctrl driver, instead of continuing extending custom pinmux or board-level pin control

I don't have time to work on a `pinctrl` driver now, and it doesn't look like anyone else is working on it yet... so to get this merged, would you consider accepting it in either state? (i.e: either merging #42512 and building off it, or the older [`saml21-no-pinctrl`](https://github.com/argentum-systems/zephyr/tree/saml21-no-pinctrl)). If "_no_", then I'm concerned this PR is going to fall away or require extensive work to keep up to date until it can be reworked and merged with this as yet non-existing / proposed `pinctrl` driver.

All of the SAM0 family will require work to shift them over, and I've done what I can to follow in their footsteps to make this as easy as possible / standardised within the family, for when the time comes.

I'll pop in to `#pinctrl` to discuss further...

gmarull commented:
So I guess this will need to wait until someone volunteers to implement a pinctrl driver for Atmel SoCs. I don't think we should make an exception when other people are doing the effort to do so.

attie-argentum commented:
This PR is currently waiting for #43731 to mature.

attie-argentum commented:
#43731 is merged, this PR is now waiting on me.

attie-argentum commented:
Hi @carlescufi / @stephanosio / @nandojve / @gmarull 

I've reworked the PR now that #43731 is merged. Please have another look over it, and let me know if there are any further issues.

gmarull commented on soc/arm/atmel_sam0/common/Kconfig.saml2x:
```diff
@@ -0,0 +1,64 @@
+config SOC_ATMEL_SAML_OSC32K
+	bool "Internal 32.768 kHz RC oscillator"
+	help
+		Say y to enable the internal 32.768 kHz RC oscillator at
```
nit: "Say y ..."  is rarely found in Zephyr

attie-argentum replied:
done

gmarull commented on soc/arm/atmel_sam0/common/soc_saml2x.c:
```diff
@@ -0,0 +1,272 @@
+	pause_for_debug();
+
+	gclk_reset();
+	osc32k_init();
+	xosc32k_init();
+	osc16m_init();
+	dfll48m_init();
```
question: why isn't this implemented as a clock control driver?

attie-argentum replied:
I can't answer that one immediately - should it be?

... I was following suit with the other SOCs

gmarull replied:
Ideally it should be, but reality is that clock control is right now a mess and there's no proper infrastructure nor guidelines in place. Not blocking, was just curious about this decision.

attie-argentum replied:
Understood, thank you.

gmarull commented on dts/arm/atmel/samr34.dtsi:
```diff
@@ -0,0 +1,60 @@
+/delete-node/ &dac;
+
+&sercom4 {
+	/* SERCOM4 will typically used to interface with the internal LoRa radio */
```
nit: `will typically be used`

attie-argentum replied:
done

gmarull commented on soc/arm/atmel_sam0/common/soc_samr34.c:
```diff
@@ -0,0 +1,34 @@
+
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/pinmux.h>
```
this header is not needed here

attie-argentum replied:
done

gmarull reviewed (commented):
Overall lgtm, just have a few comments. will let Atmel maintainers review this PR in detail.

attie-argentum commented:
Force push to address comments on zephyrproject-rtos/hal_atmel#17 and here.

nandojve commented on west.yml:
```diff
@@ -60,7 +60,7 @@ manifest:
         - hal
     - name: hal_atmel
-      revision: 78c5567c05b6b434dd7d98f49156319df4217bac
+      revision: pull/17/head
```
Hi @attie-argentum ,

You can update with new revision.

attie-argentum commented:
@gmarull - unfortunately I can't currently justify getting into implementing a clock control driver for SAM0 right now... Is this a blocker for you on this PR? If yes, we'll have to defer merging again unfortunately.

@nandojve - do you have any further comments, or are you happy with the state of this PR now?

nandojve reviewed (changes requested):
Hi @attie-argentum ,

Please, look my two recent comments and fix CI. After that I think it is ready to go.

attie-argentum commented on boards/arm/atsaml21_xpro/atsaml21_xpro-pinctrl.dtsi:
```diff
@@ -0,0 +1,66 @@
+&pinctrl {
+	tcc0_default: tcc0_default {
+		group1 {
+			pinmux =
+				<PB10F_TCC0_WO4>;
```
Referring to the line break? Sure, fixed.

nandojve replied:
It is allowed to you add spaces to match vertical alignment.
Please, do it and let`s get this in!

attie-argentum replied:
Oh, I put spaces in, but CI complained - I replaced with tabs instead... is that okay?

nandojve replied:
You must use tabs most you can than just to fix indentation you are allowed to add few spaces.

attie-argentum replied:
Ah! Understood, thanks!

attie-argentum commented on dts/arm/atmel/samr34.dtsi:
```diff
@@ -0,0 +1,60 @@
+	dipo = <0>;
+	dopo = <1>;
+	pinctrl-0 = <&pb30f_sercom4_pad2 &pb31f_sercom4_pad1
+			&pc18f_sercom4_pad3 &pc19f_sercom4_pad0>;
```
Oops, thanks - that was a mis-aimed fixup in the next commit...

attie-argentum commented:
I was going to try to look in to the failed `drivers.uart.async_api.rtt` tests again, but haven't got very far with this.
I think I got to the bottom of it last time I looked, and it was expected for them to fail, but don't remember what the reasoning was.
@nandojve - do you expect them to pass?

nandojve commented:
> I was going to try to look in to the failed drivers.uart.async_api.rtt tests again, but haven't got very far with this.
I think I got to the bottom of it last time I looked, and it was expected for them to fail, but don't remember what the reasoning was.
@nandojve - do you expect them to pass?

As far I remember, and I could be wrong, async should work DMA should be enabled. If that is the case, you need add DMA nodes for the sample on all USART that are enabled. This way the drive should build correctly. These seems to be the case that DMA has been enabled but there are no devicetree nodes and than you can get those type of messages.

nandojve commented:
Hi @attie-argentum ,

Could you add **atsaml21_xpro** and **atsamr34_xpro** at platform_exclude in the **uart_async_api** test ?

attie-argentum commented:
> Could you add **atsaml21_xpro** and **atsamr34_xpro** at platform_exclude in the **uart_async_api** test ?

Good call - others in the family are listed there too. Thanks for pointing it out!

attie-argentum commented:
Looks like the only failures are now `lcd_cyclonev_socdk` and socket tests... neither of which are relevant to this PR...

attie-argentum commented:
There will be a change required after #45703 is merged

nandojve commented:
> There will be a change required after https://github.com/zephyrproject-rtos/zephyr/pull/45703 is merged ... or we can merge this first, and put that change there instead.
(see soc/arm/atmel_sam0/common/adc_fixup_saml21.h)

This PR can go forward only when next merged window be open. I think #45703 should be merged for the current release.

attie-argentum commented:
> There will be a change required after #45703 is merged ... or we can merge this first, and put that change there instead. (see `soc/arm/atmel_sam0/common/adc_fixup_saml21.h`)

This change is now in place.

nandojve commented:
Could you rebase?

attie-argentum commented:
Yes, hopefully tomorrow / Thursday!

attie-argentum commented:
@nandojve - rebase done, no changes required.

nandojve commented on soc/arm/atmel_sam0/common/soc_saml2x.c:
```diff
@@ -0,0 +1,272 @@
+
+	key = irq_lock();
+
+	pause_for_debug();
```
My feeling is this should be at least at Kconfig. It could be dropped when you produce binary in release mode too.

attie-argentum replied:
Well caught... I was actually going to ask if anyone else has seen this sort of behaviour. (I guess not by its absence in other SoCs...)

I'll make it a config option, though to if you agree, I'll make this `default y`, otherwise people will have a terrible time getting things up and running.

attie-argentum replied:
See [here](https://github.com/zephyrproject-rtos/zephyr/pull/41308/commits/e9548d55363040712af337a088c7a087b67d6c3d#diff-1630f78dc46dcfb3dc029e1d30e8e2e35826e0a4b49dcec575facb376a0c6f90R6-R16) for commentary.

nandojve commented on dts/arm/atmel/samr34.dtsi:
```diff
@@ -0,0 +1,58 @@
+		compatible = "semtech,sx1276";
+		reg = <0>;
+		label = "SX1276";
+		reset-gpios = <&portb 15 GPIO_ACTIVE_LOW>;           /* nRST */
+		dio-gpios =
+			<&portb 16 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>, /* DIO0 */
```
Fix comment alignment.

nandojve commented on tests/drivers/uart/uart_async_api/boards/atsamr34_xpro.overlay:
```diff
@@ -0,0 +1,41 @@
+&pinctrl {
+	sercom2_default: sercom2_default {
+		group1 {
+			pinmux =
+				<PA8D_SERCOM2_PAD0>;
```
Keep at same line

nandojve commented on soc/arm/atmel_sam0/common/Kconfig.saml2x:
```diff
@@ -0,0 +1,74 @@
+
+config SOC_ATMEL_SAML_DEBUG_PAUSE
+	bool "Insert a pause at boot, to allow a debugger to attach"
+	default y
```
Remember that is always desired when you are debugging code.

```diff
-default y
+default y if DEBUG
```

attie-argentum replied:
As above, I'm not sure if this is sensible... are new users expected to know to set `CONFIG_DEBUG=y`? If not, then they won't be able to attach a debugger at all without enabling one of these options. Perhaps we should discuss?

nandojve replied:
I was wondering why you need attached a debugger if you don`t have debugger symbols. If you are building a release image this property should not be used, unless, explicit selected on my view. As far I remember, if you have reset signal you can always connect a debugger. However, if that pin is not available you may suffer problems when the code use some LPM strategy (stopped clocks and no RST pin). Is that your case? Or is there any IC BUG that you know?

attie-argentum replied:
Running `ninja debug` (or equivalent) with `CONFIG_DEBUG=n` still works - breakpoints, stepping, etc... There are no symbols in the `zephyr.bin` / `zephyr.hex` files, but the corresponding `zephyr.elf` is still useful in this case. ["_Right now, this option only disables optimization_"](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/debug/Kconfig#L95-L100)

My current understanding of the issue, is that the debugger (`gdb` + Atmel ICE) will reset the processor, and there is then a race condition between A) the processor configuring its clocks as part of the Zephyr boot, and B) the debugger halting it to do stuff... Deferring the clock configuration makes this reliable (which isn't a particularly nice "_solution_").

Erasing and/or flashing with `edbg` + Atmel ICE is 100% reliable.

I don't believe there is an IC bug, but I may be wrong.

nandojve replied:
CC: @stephanosio , @mnkp any tip about this race condition at initialization when try attach debug? It will be valuable a second opinion.

nandojve commented on soc/arm/atmel_sam0/common/soc_saml2x.c:
```diff
@@ -0,0 +1,276 @@
+	for (uint32_t i = 0; i < 10000; i += 1) {
+		__asm__ volatile ("nop\n");
+	}
+}
+#endif
```
I think it is better if you can remove `#if ... #endif` at atmel_saml_init.

```diff
+#else
+void pause_for_debug(void)
#endif
```

nandojve reviewed (dismissed):
Hi @attie-argentum ,

I was thinking about this issue. There is no harm to keep this enabled by default e people can disable when they make all tests.

Thank you for the PR and patience.

attie-argentum commented:
Great, thanks - no problem!

mbolivar-nordic commented on dts/arm/atmel/saml2x.dtsi:
```diff
@@ -0,0 +1,239 @@
+			<0x0080A048 0x4>;
+	};
+
+	aliases {
+		port-a = &porta;
+		port-b = &portb;
+		adc-0 = &adc;
+
+		sercom-0 = &sercom0;
+		sercom-1 = &sercom1;
+		sercom-2 = &sercom2;
+		sercom-3 = &sercom3;
+		sercom-4 = &sercom4;
+		sercom-5 = &sercom5;
+
+		tc-4 = &tc4;
+	};
```
Are any of these aliases actually necessary?

Many aliases in zephyr dtsi files are copy/pasted from files that were added before zephyr v2.3. In the old days, we needed aliases because there was no support for accessing nodes by node label. So for example in zephyr v2.2, you could access devicetree data from the `sercom-0` alias, but not the `sercom0` node label. But ever since v2.3, we have DT_NODELABEL() as part of the current devicetree API, so we don't need these sort of redundant aliases anymore as far as I can tell.

attie-argentum replied:
It seems not...!

mbolivar-nordic commented on dts/bindings/clock/atmel,saml2x-gclk.yaml:
```diff
@@ -0,0 +1,18 @@
+      const: 1
+
+clock-cells:
+  - periph_ch
```
Why isn't this `periph-ch`?

attie-argentum replied:
I can't answer that one, but I'm following suit from various drivers and [`dts/bindings/clock/atmel,samd5x-gclk.yaml:18`](https://github.com/zephyrproject-rtos/zephyr/blob/88ca3aca98c0788e2c45db5c3fb9c8523293a8c0/dts/bindings/clock/atmel,samd5x-gclk.yaml#L18)

... `periph-ch` doesn't exist as a string in the code base.

mbolivar-nordic commented on soc/arm/atmel_sam0/common/Kconfig.saml2x:
```diff
@@ -0,0 +1,74 @@
+config SOC_ATMEL_SAML_DEBUG_PAUSE
+	bool "Insert a pause at boot, to allow a debugger to attach"
+	default y
+	help
+		Issues have been observed while attempting to attach a debugger.
+		These can be mitigated by inserting a small delay during the early boot
+		sequence, before the system clock is configured.
+
+		If you ever intend to attach a debugger, say y.
+		If you are confident that you will never attach a debugger to the
+		resulting binary, or require the best possible boot time, say n.
```
The formatting is wrong here. Please update this to use the following style:

```
help
<TAB><SPACE><SPACE>Issues have been observed...
````

Instead of

```
help
<TAB><TAB>Issues have been observed
```

Similar comments apply elsewhere.

I know it's weird, but it's how Linux does it, so we do the same.

mbolivar-nordic commented on dts/arm/atmel/samr34.dtsi:
```diff
@@ -0,0 +1,58 @@
+	compatible = "atmel,sam0-spi";
+	dipo = <0>;
+	dopo = <1>;
+	cs-gpios = <&portb 31 GPIO_ACTIVE_LOW>;
```
Is this really correct? This doesn't look like it belongs in a SoC .dtsi file; by definition GPIOs are exterior to the SoC itself.

I would have the same question for any other similar types of settings in this PR.

attie-argentum replied:
The SAMR34 / SAMR35 is actually a SiP (System-in-Package) - these pins are internally connected between the SAML21 silicon and the SX127x silicon, and have no other functional uses... so I believe in this case, this is indeed correct.

mbolivar-nordic commented on boards/arm/atsamr34_xpro/atsamr34_xpro.dts:
```diff
@@ -0,0 +1,122 @@
+		led0 = &led0;
+		pwm-led0 = &pwm_led0;
+		sw0 = &user_button;
+		i2c-0 = &sercom2;
```
This looks like an unnecessary alias.

mbolivar-nordic reviewed (dismissed):
Since @nandojve has approved, I'd like to help get this in, but I have a few nitpicks and questions before I give my +1.

attie-argentum commented:
Rebase & address @mbolivar-nordic's comments

carlescufi commented:
@mbolivar-nordic @nandojve please take another look

gmarull commented on boards/arm/atsamr34_xpro/atsamr34_xpro_defconfig:
```diff
@@ -0,0 +1,19 @@
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_GPIO=y
+CONFIG_GPIO_SAM0=y
+CONFIG_WATCHDOG=y
+CONFIG_WDT_SAM0=y
+CONFIG_SPI_SAM0=y
```
nit: these are not minimal

attie-argentum replied:
I've removed WDT and SPI

gmarull commented on soc/arm/atmel_sam0/common/soc_saml2x.c:
```diff
@@ -0,0 +1,276 @@
+ * @brief Atmel SAML MCU series initialization code
+ */
+
+#include <arch/cpu.h>
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <soc.h>
```
please use `<zephyr/>` prefix everywhere

attie-argentum replied:
Done... `soc.h` remains without the `zephyr/` prefix, correct?

gmarull replied:
correct, soc.h is actually in the process of being removed

gmarull commented on soc/arm/atmel_sam0/common/soc_samr34.c:
```diff
@@ -0,0 +1,33 @@
+#include <drivers/gpio.h>
+#include <soc.h>
+
+#if !(ATMEL_SAM0_DT_SERCOM_CHECK(4, atmel_sam0_spi) && CONFIG_SPI_SAM0)
+/* When the radio is not in use, it's important that #CS is set high, to avoid
+ * unexpected behavior and increased current consumption... see Chapter 10 of
+ * DS70005356C. We also hold the radio in reset.
+ */
+static int soc_pinconf_init(const struct device *dev)
+{
+	const struct device *portb = DEVICE_DT_GET(DT_NODELABEL(portb));
+
+	ARG_UNUSED(dev);
+
+	if (!device_is_ready(portb)) {
+		return -ENXIO;
+	}
+
+	gpio_pin_configure(portb, 31, GPIO_OUTPUT_HIGH);
+	gpio_pin_configure(portb, 15, GPIO_OUTPUT_LOW);
+
+	return 0;
+}
+
```
why aren't these pins described in DT?

attie-argentum replied:
These pins are set to hold the radio silicon in reset when the SPI peripheral is _not_ in use... is there a way to set GPIO pins in the DT, and if so, can we set them for this negated state?

gmarull replied:
One can do things like https://github.com/zephyrproject-rtos/zephyr/pull/43817 to avoid hardcoded numbers in code.

gmarull replied:
(note: non-blocking to get this in, though)

gmarull commented on soc/arm/atmel_sam0/common/soc_samr34.c:
```diff
@@ -0,0 +1,33 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <drivers/gpio.h>
+#include <soc.h>
```
ditto

gmarull commented on soc/arm/atmel_sam0/common/soc_samr34.c:
```diff
@@ -0,0 +1,33 @@
+	ARG_UNUSED(dev);
+
+	if (!device_is_ready(portb)) {
+		return -ENXIO;
```
`-ENODEV`

gmarull reviewed (changes requested):
a few minor nitpicks before +1

attie-argentum commented:
Address @gmarull's comments and rebase.

nandojve commented:
Hi @mbolivar-nordic , @gmarull 

I was wondering if you may revisit this PR to check if @attie-argentum addressed all your concerns.

gmarull commented on dts/arm/atmel/samr34.dtsi:
```diff
@@ -0,0 +1,58 @@
+			<&portb 17 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>, /* DIO3 */
+			<&porta 10 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>, /* DIO4 */
+			<&portb  0 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>; /* DIO5 */
+		spi-max-frequency = <10000000>;
```
```suggestion
		spi-max-frequency = <DT_FREQ_M(10)>;
```

gmarull commented on samples/drivers/adc/boards/atsaml21_xpro.overlay:
```diff
@@ -0,0 +1,25 @@
+	};
+};
+
+&adc {
+	pinctrl-0 = <&adc_default>;
+	pinctrl-names = "default";
+};
```
you need to provide channel configuration in DT. Otherwise this sample won't work, see https://github.com/zephyrproject-rtos/zephyr/pull/46110/commits/94d703cd48ae97a1c7de884937e82f7ed1a233c7

attie-argentum replied:
Looks like the other SAM0 samples are broken too... `zephyr,input-positive` is missing... I'll see about a fix separately, but I can't test it.

attie-argentum commented:
@gmarull - thanks for your comments... I'm going to hang on to the GPIO (magic numbers / device tree) for the moment, if that's okay... but I've addressed the others.

Very keen to get this merged now, I seem to be periodically playing catch-up with a moving target.

gmarull reviewed (approved):
Thanks for the patience and continued effort

attie-argentum commented:
@mbolivar-nordic - please could you have a look over this again, I'd love to get this merged and complete.

mbolivar-nordic commented:
> @mbolivar-nordic - please could you have a look over this again, I'd love to get this merged and complete.

Done; you're just down to @stephanosio 's NAK now

stephanosio reviewed (approved):
LGTM

attie-argentum commented:
Amazing, thanks all! 🥳
